api/cedar_metadata_records/serializers.py

```diff
@@ -0,0 +1,116 @@
+from api.base.utils import absolute_reverse
+
+from osf.exceptions import ValidationError
+from osf.models import CedarMetadataRecord, CedarMetadataTemplate, Guid
+
+logger = logging.getLogger(__name__)
+
+
+class TargetRelationshipField(RelationshipField):
+
+    def get_object(self, _id):
+        return Guid.load(_id)
+
+    def to_internal_value(self, data):
+        return self.get_object(data)
+
+
+class CedarMetadataTemplateRelationshipField(RelationshipField):
+
+    def get_object(self, _id):
+        return CedarMetadataTemplate.load(_id)
+
+    def to_internal_value(self, data):
+        return self.get_object(data)
+
+
+class CedarMetadataRecordsSerializer(JSONAPISerializer):
+
+    class Meta:
+        type_ = 'cedar-metadata-records'
+
+    filterable_fields = frozenset(['is_published'])
+
+    id = ser.CharField(source='_id', read_only=True)
+
+    target = RelationshipField(
+        source='guid',
+        related_view='guids:guid-detail',
+        related_view_kwargs={'guids': '<guid._id>'},
+        # always_embed=True,
+        read_only=True,
+    )
+
+    template = RelationshipField(
+        related_view='cedar-metadata-templates:cedar-metadata-template-detail',
+        related_view_kwargs={'template_id': '<template._id>'},
+        # always_embed=True,
+        read_only=True,
+    )
+
+    metadata = ser.DictField(read_only=False)
+
+    is_published = ser.BooleanField(read_only=False)
+
+    links = LinksField({'self': 'get_absolute_url'})
+
+    def get_absolute_url(self, obj):
+        return absolute_reverse('cedar-metadata-records:cedar-metadata-record-detail', kwargs={'record_id': obj._id})
+
+    def update(self, instance, validated_data):
+        assert isinstance(instance, CedarMetadataRecord), 'instance must be a CedarMetadataRecord'
+        for key, value in validated_data.items():
+            if key == 'metadata':
+                instance.metadata = value
+            elif key == 'is_published':
+                instance.is_published = value
+            else:
+                continue  # ignore other attributes
+        instance.save()
+        return instance
+
+
+class CedarMetadataRecordsCreateSerializer(CedarMetadataRecordsSerializer):
+
+    target = TargetRelationshipField(
+        source='guid',
+        related_view='guids:guid-detail',
+        related_view_kwargs={'guids': '<guid._id>'},
+        # always_embed=True,
+        read_only=False,
+        required=True,
+    )
+
+    template = CedarMetadataTemplateRelationshipField(
+        related_view='cedar-metadata-templates:cedar-metadata-template-detail',
+        related_view_kwargs={'template_id': '<template._id>'},
+        # always_embed=True,
+        read_only=False,
+        required=True,
+    )
+
+    metadata = ser.DictField(read_only=False, required=True)
+
+    is_published = ser.BooleanField(read_only=False, required=True)
+
+    def create(self, validated_data):
+
+        guid = validated_data.pop('guid')
```

Comment:
For CR: when a field in model and serializer differs, `serializer.validated_data` uses model field name `guid` instead of the serializer's `target`.

Reply:
Even if the payload from the FE says `target`?

Reply:
Yes, which is surprising to me as well. There is not many relationship fields that use different names, but I couldn't find either an example or a counter-example.

Reply:
Here is my `ipdb` log

```
ipdb> self.validated_data
{'guid': <id:mzp5j, referent:((title='Cedar Metadata - Project 0001', category='project') with guid 'mzp5j')>, 'template': <CedarMetadataTemplate: CedarMetadataTemplate object (4)>, 'metadata': {'random': 'random2', 'modnar': 'modnar2'}, 'is_published': True}
ipdb> self.initial_data
{'id': None, 'type': 'cedar_metadata_records', 'metadata': {'random': 'random2', 'modnar': 'modnar2'}, 'is_published': 'true', 'template': 'cedar0002', 'target': 'mzp5j'}
```

`initial_data` is the data from request